### PR TITLE
Fix a bug in use_original_dst option

### PR DESCRIPTION
**Pull request by kyessenov**

### CODE REVIEW

**Comment by mattklein123** [comment]
Is it possible to somehow have a test that covers this?  I'm wondering if we could somehow have a virtual function that resolves the original address so that we can than test it? I'm fine getting this fix in to unblock you but if we could follow up with a test that would be great.

**Comment by mattklein123** [comment]
I think with the virtual function and a TestListenerImpl class we could probably test it pretty easily following along here:
https://github.com/lyft/envoy/blob/master/test/common/network/listener_impl_test.cc

**Comment by kyessenov** [comment]
Added a test as you suggested. PTAL.

File: test/common/network/listener_impl_test.cc [hunk 1]
```diff
@@ -38,4 +39,60 @@ TEST(ListenerImplDeathTest, ErrorCallback) {
...
+  EXPECT_CALL(listener, getAddressPort_(_)).WillRepeatedly(Return(10001));
+  EXPECT_CALL(connection_handler, findListener("10001")).WillRepeatedly(Return(&listenerDst));
+
+  EXPECT_CALL(listener, newConnection_(_, _)).Times(0);
+  EXPECT_CALL(listenerDst, newConnection_(_, _))
+      .Times(1)
+      .WillOnce(Invoke([&](int, sockaddr*) -> void {
+        client_connection->close(ConnectionCloseType::NoFlush);
+        socket.close();
+        socketDst.close();
```
Comment (mattklein123) [inline test/common/network/listener_impl_test.cc]: These 2 lines in the above test are just kind of a hack to force the error callback to be called. (Sorry I should have a comment to that effect, perhaps you could add one).

Can we make this test a non-death test? If you just call dispatcher.exit() here the event loop will exit and the test will do what is expected I think (expect a call of newConnection on the other listener).

File: test/common/network/listener_impl_test.cc [hunk 2]
```diff
@@ -38,4 +39,60 @@ TEST(ListenerImplDeathTest, ErrorCallback) {
...
+      }));
+
+  dispatcher.run(Event::Dispatcher::RunType::Block);
+}
+
+TEST(ListenerImpl, UseOriginalDst) {
```
Comment (mattklein123) [inline test/common/network/listener_impl_test.cc]: nit: ListenerImplTest

File: test/common/network/listener_impl_test.cc (hunk 1 above)
Comment (kyessenov) [inline test/common/network/listener_impl_test.cc]: Yeah, that's better. I wasn't sure how to terminate the loop.

File: test/common/network/listener_impl_test.cc (hunk 2 above)
Comment (kyessenov) [inline test/common/network/listener_impl_test.cc]: Fixed!

**Pull request merged**